src/lib/Libnet/hnls.c

```diff
@@ -574,53 +577,165 @@ free_if_info(struct log_net_info *ni)
+			tmp = realloc(nodenames, len + hlen + 2); /* 2 for comma and null char */
+			if (!tmp) {
+				strncpy(msg_buf, "Out of memory", msg_buf_len);
+				free(nodenames);
+				nodenames = NULL;
+				break;
+			}
+			nodenames = tmp;
+
+			if (len == 0)
+				strcpy(nodenames, buf);
+			else {
+				strcat(nodenames, ",");
+				strcat(nodenames, buf);
+			}
+			len += hlen + 2;
+			count++;
+		}
+	}
+
+	freeaddrinfo(pai);
+
+	if (count == 0) {
+		snprintf(msg_buf, msg_buf_len, "Could not find any usable IP address for host %s", host);
+		return NULL;
+	}
+	return nodenames;
+}
+
+/**
+* @brief
+*	Get a list of all IPs
+*   First it resolves the supplied hostname to determine it's IPs
+*	Then it enumerates the interfaces in the host and determines IPs
+*	for each of those interfaces.
+*
+*	Do not supply a remote hostname in this function.
+*
+* @return
+*	Comma separated list of IPs in string format
 *
-* @return 
-*	Comma separated list of ips in string format
-* 
 * @par Side Effects:
 *	None
 *
 * @par MT-safe: Yes
 *
-* @param[out]   msg - error message returned if system calls not successful
+* @param[in]	hostname    - hostname of the current host to resolve IPs for
+* @param[out]   msg_buf     - error message returned if system calls not successful
+* @param[in]    msg_buf_len - length of the message buffer passed
 *
 */
 char *
-get_all_ips(char *msg)
+get_all_ips(char *hostname, char *msg_buf, size_t msg_buf_len)
 {
-	char * nodenames = NULL;
-#if defined(linux)
+	char *nodenames;
+	int len, ret;
 	char *tmp;
-	int len, hlen, ret;
+	char buf[NETADDR_BUF];
+
+#if defined(linux)
 	struct ifaddrs *ifp, *listp;
-	char buf[NETADDR_BUF] = {'\0'};
 	char *p;
+#elif defined(WIN32)
+	int i;
+	/* Variables used by GetIpAddrTable */
+	PMIB_IPADDRTABLE pIPAddrTable;
+	DWORD dwSize = 0;
+	DWORD dwRetVal = 0;
+	IN_ADDR IPAddr;
+#endif
+
+	msg_buf[0] = '\0';
+
+	/* prepend the list of IPs with the IPs resolved from the passed hostname */
+	nodenames = get_host_ips(hostname, msg_buf, msg_buf_len);
+	if (!nodenames) {
+		return NULL;
+	}
 
-	msg[0] = '\0';
+	len = strlen(nodenames);
 
+#if defined(linux)
 	ret = getifaddrs(&ifp);
 
 	if ((ret != 0) || (ifp == NULL)) {
-		strncpy(msg, "Failed to obtain interface names", LOG_BUF_SIZE - 1);
+		strncpy(msg_buf, "Failed to obtain interface names", msg_buf_len);
+		free(nodenames);
 		return NULL;
 	}
 
-	len = 0;
 	for (listp = ifp; listp; listp = listp->ifa_next) {
+		int hlen;
+
 		if ((listp->ifa_addr == NULL) || (listp->ifa_addr->sa_family != AF_INET)) 
 			continue;
 		sprintf(buf, "%s", netaddr((struct sockaddr_in *)listp->ifa_addr));
-		if (!strcmp(buf,"unknown")) 
+		if (!strcmp(buf,"unknown"))
 			continue;
-		if ((p=strchr(buf, ':')))
+		if ((p = strchr(buf, ':')))
 			*p = '\0';
 
 		hlen = strlen (buf);
 		tmp = realloc(nodenames, len + hlen + 2); /* 2 for comma and null char */
 		if (!tmp) {
-			strncpy(msg, "Out of memory", LOG_BUF_SIZE - 1);
+			strncpy(msg_buf, "Out of memory", msg_buf_len);
```

Comment:
Not that it matters much but seems like like 745 and 746 is not needed because you already check for nodenames to be non-null. Same comment will apply to windows code as well.
I'm ok if you choose not to change it

Reply:
Ya, I see your point. We can remove it, but then nodenames can be non-null but strlen (ie, len) could be 0, it is technically it could be an empty string- the likelihood of that is very rare, but given that I will keep the code as is. Thanks